File: src/main/java/aurocosh/divinefavor/common/item/talismans/base/TalismanContext.java
package aurocosh.divinefavor.common.item.talismans.base;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class TalismanContext {
    public EntityPlayer player;
    public World world;
    public BlockPos pos;
    public EnumHand hand;
    public EnumFacing facing;
    public CastType castType;

    public TalismanContext(EntityPlayer player) {
        this.player = player;
        this.world = player.world;
        this.pos = null;
        this.hand = EnumHand.MAIN_HAND;
        castType = CastType.UNSPECIFIED;
    }

    public TalismanContext(EntityPlayer player, World world, BlockPos pos, EnumHand hand, EnumFacing facing, CastType castType){
        this.player = player;
        this.world = world;
        this.pos = pos;
        this.hand = hand;
        this.facing = facing;
        this.castType = castType;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/spell_talismans/context/TalismanContext.kt
package aurocosh.divinefavor.common.item.spell_talismans.context

import aurocosh.divinefavor.common.item.spell_talismans.base.CastType
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d
import net.minecraft.world.World

data class ContextProperty<T>(val name: String, val defaultValue: T)
data class CommonContextValues(val player: EntityPlayer, val stack: ItemStack, val world: World)

class TalismanContext(
        val player: EntityPlayer, val target: EntityLivingBase?, val world: World,
        val pos: BlockPos, val posVec: Vec3d, val hand: EnumHand, val facing: EnumFacing,
        val castType: CastType, val stack: ItemStack,
        val raycastValid: Boolean, val stackValid: Boolean) {

    private val properties = HashMap<String, Any>()

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> get(property: ContextProperty<T>): T {
        val value = (properties[property.name] ?: property.defaultValue)
        return value as T
    }

    fun <T : Any> set(property: ContextProperty<T>, value: T) {
        properties[property.name] = value
    }

    fun getCommon() = CommonContextValues(player, stack, world)
}
